### DISCUSSION TITLE
Concurrent request problem

### DISCUSSION THREAD

**Discussion by SCParrot** [mitmproxy/mitmproxy#7180] [Q&A] [answered]
```python
class CffiAddon:
    def request(self, flow):
        request = flow.request

        time.sleep(5)
```
When I use the above code to process 5 concurrent requests, I wait 5s to send a request and then wait 5s to send another request, instead of sending 5 requests at once and waiting 5s.
Is there any way to process this project concurrently?

**Comment by emanuele-em** [comment, marked as answer]
I believe you might find _asyncio_ helpful! Check out this [example](https://docs.mitmproxy.org/stable/addons-examples/#nonblocking) in the documentation—it seems to be exactly what you're looking for 🙂

**Reply by SCParrot** [reply]
thanks, it really help me
